### PR TITLE
Create separate reconciler for otel-collector

### PR DESCRIPTION
**Pull request by andriisoldatenko**
## Description

This PR extracts and move pkg/controllers/dynakube/extension/otel to pkg/controllers/dynakube/otelc and prepares it for future featues.

resolves https://dt-rnd.atlassian.net/browse/DAQ-3315

### motivation

In the future, we will use OtelCollector Statefulset for a few features and not only extensions, so it deserves a separate small reconciler. Alos OtelCollector can be configured using dedicated new section (already exists) `dk.Spec.Templates.OpenTelemetryCollector`, so we don't have conflict of interests of different features, it's just shared collector with appended config (in future) as additional arguments to OtelC container.

## How can this be tested?

should work as before refactoring

```
make test/e2e/extensions
```

### CODE REVIEW

File: pkg/controllers/dynakube/otelc/statefulset/env.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	envShardId            = "SHARD_ID"
+	envPodNamePrefix      = "POD_NAME_PREFIX"
+	envPodName            = "POD_NAME"
+	envOTLPgrpcPort       = "OTLP_GRPC_PORT"
+	envOTLPhttpPort       = "OTLP_HTTP_PORT"
+	envOTLPtoken          = "OTLP_TOKEN"
+	envEECDStoken         = "EEC_DS_TOKEN"
+	envTrustedCAs         = "TRUSTED_CAS"
+	envK8sClusterName     = "K8S_CLUSTER_NAME"
+	envK8sClusterUuid     = "K8S_CLUSTER_UID"
```
Comment (waodim) [inline pkg/controllers/dynakube/otelc/statefulset/env.go:28]:
```suggestion
	envK8sClusterUid     = "K8S_CLUSTER_UID"
```

File: pkg/controllers/dynakube/otelc/config.go [hunk 2]
```diff
@@ -1,9 +1,9 @@
-package otel
+package otelc
 
 import (
 	"github.com/Dynatrace/dynatrace-operator/pkg/logd"
 )
 
 var (
-	log = logd.Get().WithName("extension-otel")
+	log = logd.Get().WithName("otelc")
```
Comment (waodim) [inline pkg/controllers/dynakube/otelc/config.go:8]: wdyt about renaming to otel-collector? or do you think it is too long?

**Review by waodim — COMMENTED** [review]
just some nitpicks - will give it a go later and test

File: pkg/controllers/dynakube/otelc/config.go (hunk 2 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/otelc/config.go:8]: it's good idea will do :)

**Review by aorcholski — DISMISSED** [review]
LGTM
e2e PASS
a few manual tests and it also works as always

File: pkg/controllers/dynakube/controller.go [hunk 3]
```diff
@@ -332,6 +335,17 @@ func (controller *Controller) reconcileComponents(ctx context.Context, dynatrace
 		componentErrors = append(componentErrors, err)
 	}
 
+	log.Info("start reconciling app otelc")
```
Comment (0sewa0) [inline pkg/controllers/dynakube/controller.go:338]: does the `app` part needed? 😅

File: pkg/controllers/dynakube/extension/consts/consts.go [hunk 4]
```diff
@@ -1,33 +1,14 @@
 package consts
 
-import "github.com/Dynatrace/dynatrace-operator/pkg/api"
-
 const (
-	ExtensionsAnnotationSecretHash = api.InternalFlagPrefix + "secret-hash"
+	// TODO rename without extra prefix use import alias
```
Comment (0sewa0) [inline pkg/controllers/dynakube/extension/consts/consts.go:4]: ?

File: pkg/controllers/dynakube/extension/consts/consts.go [hunk 5]
```diff
@@ -1,33 +1,14 @@
...
-
-	ExtensionsSelfSignedTLSSecretSuffix     = "-extensions-controller-tls"
-	ExtensionsSelfSignedTLSCommonNameSuffix = "-extensions-controller.dynatrace"
+	TokenSecretKey         = "eec.token"
+	TokenSecretValuePrefix = "EEC dt0x01"
 
-	// TLSKeyDataName is the key used to store a TLS private key in the secret's data field.
-	TLSKeyDataName = "tls.key"
+	SecretConditionType  = "ExtensionsSecret"
+	ServiceConditionType = "ExtensionsService"
```
Comment (0sewa0) [inline pkg/controllers/dynakube/extension/consts/consts.go:11]: do these need to be public?
can't they just "closer" where they are used?

File: pkg/controllers/dynakube/otelc/statefulset/reconciler.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+func buildAppLabels(dkName string) *labels.AppLabels {
+	// TODO: when version is available
+	version := "0.0.0"
+
+	return labels.NewAppLabels(labels.CollectorComponentLabel, dkName, labels.CollectorComponentLabel, version)
+}
+
+func buildAffinity() corev1.Affinity {
+	// TODO: implement new attributes in CR dk.Spec.Templates.OpenTelemetryCollector.Affinity
+	// otherwise to use defaults ones
```
Comment (0sewa0) [inline pkg/controllers/dynakube/otelc/statefulset/reconciler.go:178]: ?

File: pkg/util/kubeobjects/topology/constraint.go [hunk 7]
```diff
@@ -1,13 +1,12 @@
...
 
 import (
 	"github.com/Dynatrace/dynatrace-operator/pkg/util/kubeobjects/labels"
-	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-func BuildTopologySpreadConstraints(topologySpreadConstraints []corev1.TopologySpreadConstraint, appLabels *labels.AppLabels) []corev1.TopologySpreadConstraint {
+func SpreadConstraints(topologySpreadConstraints []corev1.TopologySpreadConstraint, appLabels *labels.AppLabels) []corev1.TopologySpreadConstraint {
```
Comment (0sewa0) [inline pkg/util/kubeobjects/topology/constraint.go:9]: you moved it to a very general place
now the name of it is misleading

It is a "use provided or set defaults"

I would:
- move the `if len(topologySpreadConstraints) > 0` check where you use this
  - having something to overwrite it is a component specific things, you are in a non-component-specific package
- have the rest remain here and rename the func to:
```
func MaxOnePerNode(appLabels *labels.AppLabels) []corev1.TopologySpreadConstraint {
return []corev1.TopologySpreadConstraint{
			{
				MaxSkew:           1,
				TopologyKey:       "topology.kubernetes.io/zone",
				WhenUnsatisfiable: "ScheduleAnyway",
				LabelSelector:     &metav1.LabelSelector{MatchLabels: appLabels.BuildMatchLabels()},
			},
			{
				MaxSkew:           1,
				TopologyKey:       "kubernetes.io/hostname",
				WhenUnsatisfiable: "DoNotSchedule",
				LabelSelector:     &metav1.LabelSelector{MatchLabels: appLabels.BuildMatchLabels()},
			},
		}
}
```

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: pkg/controllers/dynakube/otelc/statefulset/env.go (hunk 1 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/otelc/statefulset/env.go]: fixed in 1e7389728

File: pkg/controllers/dynakube/otelc/config.go (hunk 2 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/otelc/config.go]: fixed in 1e7389728

File: pkg/controllers/dynakube/controller.go (hunk 3 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/controller.go]: fixed in 1e7389728

File: pkg/controllers/dynakube/extension/consts/consts.go (hunk 4 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/extension/consts/consts.go]: fixed in 1e7389728

File: pkg/controllers/dynakube/extension/consts/consts.go (hunk 5 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/extension/consts/consts.go]: fixed in 1e7389728

File: pkg/controllers/dynakube/otelc/statefulset/reconciler.go (hunk 6 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/otelc/statefulset/reconciler.go]: fixed in 1e7389728

File: pkg/util/kubeobjects/topology/constraint.go (hunk 7 above)
Comment (andriisoldatenko) [reply pkg/util/kubeobjects/topology/constraint.go]: fixed in 1e7389728

**Review by aorcholski — APPROVED** [review]
I repeated the tests.

**Review by waodim — APPROVED** [review]

**Pull request merged**